sources/web/datalab/polymer/components/datalab-files/datalab-files.ts

```diff
@@ -258,7 +262,163 @@ class FilesElement extends Polymer.Element {
     this.$.forwardNav.disabled = this._pathHistoryIndex === this._pathHistory.length - 1;
   }
 
+  _createNewNotebook() {
+    this._createNewItem('notebook');
+  }
+
+  _createNewDirectory() {
+    this._createNewItem('directory');
+  }
+
+  /**
+   * The Jupyter contents API does not provide a way to create a new item with a specific
+   * name, it only creates new untitled files or directories in provided path (see
+   * https://github.com/jupyter/notebook/blob/master/notebook/services/contents/manager.py#L311).
+   * In order to offer a better experience, we create an untitled item in the current path,
+   * then rename it to whatever the user specified.
+   * 
+   * This method creates an input modal to get the user input, then calls the ApiManager to
+   * create a new notebook/directory at the current path, and fetches the updated list
+   * of files to redraw.
+   */
+  _createNewItem(type: string) {
+
+    // First, open a dialog to let the user specify a name for the notebook.
+    const inputOptions: DialogOptions = {
+      title: 'New ' + type, 
+      withInput: true,
+      inputLabel: 'Name',
+      okTitle: 'Create',
+    };
+
+    return Utils.getUserInputAsync(inputOptions)
+      .then((closeResult: DialogCloseResult) => {
+        // Only if the dialog has been confirmed with some user input, rename the
+        // newly created file. Then if that is successful, reload the file list
+        if (closeResult.confirmed && closeResult.userInput) {
+          let newName = this.currentPath + '/' + closeResult.userInput;
+          // Make sure the name ends with .ipynb for notebooks for convenience
+          if (type === 'notebook' && !newName.endsWith('.ipynb')) {
+            newName += '.ipynb';
+          }
+          let notebookPlaceholderName = '';
+          return ApiManager.createNewItem(type)
+            .then((notebook: JupyterFile) => {
+              notebookPlaceholderName = notebook.path;
+              return ApiManager.renameItem(notebookPlaceholderName, newName);
+            })
+            .then(this._fetchFileList.bind(this), (e: number) => {
```

Comment:
Rather than bind, might be better to use an arrow function, as you do on line 351.